File: SACSegmentation.lua
local ffi = require 'ffi'
local torch = require 'torch'
local utils = require 'pcl.utils'
local pcl = require 'pcl.PointTypes'

local SACSegmentation = torch.class('pcl.SACSegmentation', pcl)
local SACSegmentationFromNormals, SACSegmentation = torch.class('pcl.SACSegmentationFromNormals', 'pcl.SACSegmentation', pcl)

local SACSegmentation_func_by_type = {}
local SACSegmentationFromNormals_func_by_type = {}

local function init()
  local SACSegmentation_method_names = {
    'new',
    'delete',
    'SACSegmentation_ptr',
    'setInputCloud',
    'setIndices',
    'setModelType',
    'getModelType',
    'setDistanceThreshold',
    'getDistanceThreshold',
    'setMaxIterations',
    'getMaxIterations',
    'setProbability',
    'getProbability',
    'setOptimizeCoefficients',
    'getOptimizeCoefficients',
    'setSamplesMaxDist_KdTree',
    'setSamplesMaxDist_Octree',
    'setRadiusLimits',
    'setAxis',
    'getAxis',
    'setEpsAngle',
    'getEpsAngle',
    'segment'
  }
  for k,v in pairs(utils.type_key_map) do
    SACSegmentation_func_by_type[k] = utils.create_typed_methods("pcl_SACSegmentation_TYPE_KEY_", SACSegmentation_method_names, v)
  end
  
  local SACSegmentationFromNormals_method_names = {
    'new',
    'delete',
    'SACSegmentationFromNormals_ptr',
    'SACSegmentation_ptr'
  }
  for k,v in pairs(utils.type_key_map) do
    SACSegmentationFromNormals_func_by_type[k] = utils.create_typed_methods("pcl_SACSegmentationFromNormals_TYPE_KEY_", SACSegmentationFromNormals_method_names, v)
  end
end

init()

function SACSegmentation:__init(pointType)
  self.pointType = pcl.pointType(pointType or pcl.PointXYZ)
  self.f = SACSegmentation_func_by_type[self.pointType]
  self.h = self.f.new()
  self.p = self.f.SACSegmentation_ptr(self.h)
end

function SACSegmentation:handle()
  return self.h
end

function SACSegmentation:SACSegmentation_ptr()
  return self.p
end

function SACSegmentation:setInputCloud(cloud)
  self.f.setInputCloud(self.p, cloud:cdata())
end

function SACSegmentation:setIndices(indices)
  self.f.setIndices(self.p, indices:cdata())
end

function SACSegmentation:setModelType(model)
  self.f.setModelType(self.p, model)
end

function SACSegmentation:getModelType()
  return self.f.getModelType(self.p)
end

function SACSegmentation:setDistanceThreshold()
  self.f.setDistanceThreshold(self.p, threshold)
end

function SACSegmentation:getDistanceThreshold()
  return self.f.getDistanceThreshold(self.p)
end

function SACSegmentation:setMaxIterations(max_iterations)
  self.f.setMaxIterations(self.p, max_iterations)
end

function SACSegmentation:getMaxIterations()
  return self.f.getMaxIterations(self.p)
end

function SACSegmentation:setProbability(probability)
  self.f.setProbability(self.p, probability)
end

function SACSegmentation:getProbability()
  return self.f.getProbability(self.p)
end

function SACSegmentation:setOptimizeCoefficients(optimize)
  self.f.setOptimizeCoefficients(self.p, optimize)
end

function SACSegmentation:getOptimizeCoefficients()
  return self.f.getOptimizeCoefficients(self.p)
end

function SACSegmentation:setSamplesMaxDist(radius, search)
  if torch.isTypeOf(search, pcl.KdTree) then
    self.f.setSamplesMaxDist_KdTree(self.p, radius, search:cdata())
  elseif torch.isTypeOf(search, pcl.Octree) then
    self.f.setSamplesMaxDist_Octree(self.p, radius, search:cdata())
  else
    error("unsupported search method")
  end
end

function SACSegmentation:setRadiusLimits(min_radius, max_radius)
  self.f.setRadiusLimits(self.p, min_radius, max_radius)
end

function SACSegmentation:setAxis(axis)
  self.f.setAxis(self.p, axis:cdata())
end

function SACSegmentation:getAxis(result)
  result = result or torch.FloatTensor()
  self.f.getAxis(self.p, result:cdata())
  return result
end

function SACSegmentation:setEpsAngle(ea)
  self.f.setEpsAngle(self.p, ea)
end

function SACSegmentation:getEpsAngle()
  return self.f.getEpsAngle(self.p)
end

function SACSegmentation:segment(inliers, coefficients)
  inliers = inliers or pcl.Indices()
  coefficients = coefficients or torch.FloatTensor()  
  self.f.segment(self.p, inliers:cdata(), coefficients:cdata())
end

--[[
  SACSegmentationFromNormals
]]
function SACSegmentationFromNormals:__init(pointType)
  self.pointType = pcl.pointType(pointType or pcl.PointXYZ)
  self.f = SACSegmentation_func_by_type[self.pointType]
  self.f2 = SACSegmentationFromNormals_func_by_type[self.pointType]
  self.h = self.f2.new()
  self.p = self.f2.SACSegmentation_ptr(self.h)
  self.pn = self.f2.SACSegmentationFromNormals_ptr(self.h)
end

function SACSegmentationFromNormals:handle()
  return self.h
end

function SACSegmentationFromNormals:SACSegmentationFromNormals_ptr()
  return self.pn
end

function SACSegmentationFromNormals:setInputNormals(normals)
  self.f2.setInputNormals(self.pn, normals:cdata())
end

function SACSegmentationFromNormals:setNormalDistanceWeight(distance_weight)
  self.f2.setNormalDistanceWeight(self.pn, distance_weight)
end

function SACSegmentationFromNormals:setMinMaxOpeningAngle(min_angle, max_angle)
  self.f2.setMinMaxOpeningAngle(self.pn, min_angle, max_angle)
end

function SACSegmentationFromNormals:setDistanceFromOrigin(d)
  self.f2.setDistanceFromOrigin(self.pn, d)
end

File: src/generic/sac_segmentation.cpp
#include "searchwrapper.h"
#include <pcl/segmentation/sac_segmentation.h>

#define SACSegmentation_ pcl::SACSegmentation<_PointT>
#define SACSegmentation_handle boost::shared_ptr<SACSegmentation_>
#define SACSegmentationFromNormals_ pcl::SACSegmentationFromNormals<_PointT, pcl::Normal>
#define SACSegmentationFromNormals_handle boost::shared_ptr<SACSegmentationFromNormals_>

#define PointCloud_ptr pcl::PointCloud<_PointT>::Ptr
#define KdTree_ptr pcl::KdTreeFLANN<_PointT>::Ptr
#define Octree_ptr pcl::octree::OctreePointCloudSearch<_PointT>::Ptr

PCLIMP(SACSegmentation_handle *, SACSegmentation, new)()
{
  return new SACSegmentation_handle(new pcl::SACSegmentation<_PointT>());
}

PCLIMP(void, SACSegmentation, delete)(SACSegmentation_handle *handle)
{
  delete handle;
}

PCLIMP(SACSegmentation_ *, SACSegmentation, SACSegmentation_ptr)(SACSegmentation_handle *handle)
{
  return handle->get();
}

// PCLBase methods
PCLIMP(void, SACSegmentation, setInputCloud)(SACSegmentation_ *self, PointCloud_ptr *cloud)
{
  self->setInputCloud(*cloud);
}

PCLIMP(void, SACSegmentation, setIndices)(SACSegmentation_ *self, Indices_ptr *indices)
{
  self->setIndices(*indices);
}

// SACSegmentation methods
PCLIMP(void, SACSegmentation, setModelType)(SACSegmentation_ *self, int model)
{
 self->setModelType(model);
}

PCLIMP(int, SACSegmentation, getModelType)(SACSegmentation_ *self) 
{
  return self->getModelType();
}

PCLIMP(void, SACSegmentation, setDistanceThreshold)(SACSegmentation_ *self, double threshold)
{
  self->setDistanceThreshold(threshold);
}

PCLIMP(double, SACSegmentation, getDistanceThreshold)(SACSegmentation_ *self)
{
  return self->getDistanceThreshold();
}

PCLIMP(void, SACSegmentation, setMaxIterations)(SACSegmentation_ *self, int max_iterations)
{
  self->setMaxIterations(max_iterations);
}

PCLIMP(int, SACSegmentation, getMaxIterations)(SACSegmentation_ *self)
{
  return self->getMaxIterations();
}

PCLIMP(void, SACSegmentation, setProbability)(SACSegmentation_ *self, double probability)
{
  self->setProbability(probability);
}

PCLIMP(double, SACSegmentation, getProbability)(SACSegmentation_ *self)
{
  return self->getProbability();
}

PCLIMP(void, SACSegmentation, setOptimizeCoefficients)(SACSegmentation_ *self, bool optimize)
{
  self->setOptimizeCoefficients(optimize);
}

PCLIMP(bool, SACSegmentation, getOptimizeCoefficients)(SACSegmentation_ *self)
{
  return self->getOptimizeCoefficients();
}

PCLIMP(void, SACSegmentation, setSamplesMaxDist_KdTree)(SACSegmentation_ *self, double radius, KdTree_ptr *search)
{
  pcl::search::Search<_PointT>::Ptr search_(new KdTreeSearchWrapper<_PointT>(*search));
  self->setSamplesMaxDist(radius, search_);
}

PCLIMP(void, SACSegmentation, setSamplesMaxDist_Octree)(SACSegmentation_ *self, double radius, Octree_ptr *search)
{
  pcl::search::Search<_PointT>::Ptr search_(new OctreeSearchWrapper<_PointT>(*search));
  self->setSamplesMaxDist(radius, search_);
}

PCLIMP(void, SACSegmentation, setRadiusLimits)(SACSegmentation_ *self, double min_radius, double max_radius)
{
  self->setRadiusLimits(min_radius, max_radius);
}

PCLIMP(void, SACSegmentation, setAxis)(SACSegmentation_ *self, THFloatTensor *axis) 
{
  self->setAxis(Tensor2Vec3f(axis));
}

PCLIMP(void, SACSegmentation, getAxis)(SACSegmentation_ *self, THFloatTensor *result) 
{
  Eigen::Vector3f axis = self->getAxis();
  copyMatrix(axis, result);
}

PCLIMP(void, SACSegmentation, setEpsAngle)(SACSegmentation_ *self, double ea)
{
  self->setEpsAngle(ea);
}

PCLIMP(double, SACSegmentation, getEpsAngle)(SACSegmentation_ *self)
{
  return self->getEpsAngle();
}

PCLIMP(void, SACSegmentation, segment)(SACSegmentation_ *self, Indices_ptr *inliers, THFloatTensor *model_coefficients)
{
  pcl::ModelCoefficients model_coefficients_;
  pcl::PointIndices inliers_;
  self->segment(inliers_, model_coefficients_);
  if (inliers != NULL)
    **inliers = inliers_.indices;
  vector2Tensor(model_coefficients_.values, model_coefficients);
}

// SACSegmentationFromNormals
PCLIMP(SACSegmentationFromNormals_handle *, SACSegmentationFromNormals, new)()
{
  return new SACSegmentationFromNormals_handle(new pcl::SACSegmentationFromNormals<_PointT, pcl::Normal>());
}

PCLIMP(void, SACSegmentationFromNormals, delete)(SACSegmentationFromNormals_handle *handle)
{
  delete handle;
}

PCLIMP(SACSegmentationFromNormals_ *, SACSegmentationFromNormals, SACSegmentationFromNormals_ptr)(SACSegmentationFromNormals_handle *handle)
{
  return handle->get();
}

PCLIMP(SACSegmentation_ *, SACSegmentationFromNormals, SACSegmentation_ptr)(SACSegmentationFromNormals_handle *handle)
{
  return static_cast<SACSegmentation_ *>(handle->get());
}

PCLIMP(void, SACSegmentationFromNormals, setInputNormals)(SACSegmentationFromNormals_ *self, Normals_ptr *normals)
{
  self->setInputNormals(*normals);
}

PCLIMP(void, SACSegmentationFromNormals, setNormalDistanceWeight)(SACSegmentationFromNormals_ *self, double distance_weight)
{
  self->setNormalDistanceWeight(distance_weight);
}

PCLIMP(void, SACSegmentationFromNormals, setMinMaxOpeningAngle)(SACSegmentationFromNormals_ *self, double min_angle, double max_angle)
{
  self->setMinMaxOpeningAngle(min_angle, max_angle);
}

PCLIMP(void, SACSegmentationFromNormals, setDistanceFromOrigin)(SACSegmentationFromNormals_ *self, const double d)
{
  self->setDistanceFromOrigin(d);
}

#undef SACSegmentation_ 
#undef SACSegmentation_handle
#undef SACSegmentationFromNormals_
#undef SACSegmentationFromNormals_handle
#undef PointCloud_ptr
#undef KdTree_ptr
#undef Octree_ptr
